src/utils/useActivate.ts

```diff
@@ -0,0 +1,102 @@
+import { ipcRenderer } from "electron";
+import { GraphQLClient } from "graphql-request";
+import { NodeInfo } from "src/config";
+import { getSdk } from "src/generated/graphql-request";
+import { ActivationFunction, ActivationStep } from "src/interfaces/activation";
+import { useTx } from "src/utils/useTx";
+import { useLoginSession } from "./useLoginSession";
+import { useStore } from "./useStore";
+
+export function useActivate(): ActivationFunction {
+  const tx = useTx();
+  const { address, publicKey } = useLoginSession();
+  const { activationKey } = useStore("account");
+
+  const activate: ActivationFunction = async () => {
+    let step: ActivationStep = "getActivationInfo";
+
+    if (!(address && publicKey)) {
+      return {
+        result: false,
+        error: {
+          error: new Error("No address"),
+          step,
+        },
+      };
+    }
+
+    if (!activationKey) {
+      return {
+        result: false,
+        error: {
+          error: new Error("No activation key"),
+          step,
+        },
+      };
+    }
+
+    try {
+      step = "getGraphQLClient";
+      const nodeInfo: NodeInfo = await ipcRenderer.invoke("get-node-info");
+
+      const sdks = getSdk(
+        new GraphQLClient(
+          `http://${nodeInfo.host}:${nodeInfo.graphqlPort}/graphql`
+        )
+      );
+
+      step = "getActivationAddress";
+      const { data: activationData } = await sdks.ActivationAddress({
+        address,
+      });
+
+      if (activationData.activationStatus.addressActivated) {
+        return {
+          result: false,
+          error: {
+            error: new Error("Address Already activated"),
+            step,
+          },
+        };
+      }
+
+      step = "createActivateAccountTx";
+      const { data: activateData } = await sdks.ActivateAccount({
+        activationCode: activationKey,
+        publicKey,
+      });
+
+      step = "stageTx";
+      const { data: txData } = await tx(
+        activateData.actionTxQuery.activateAccount
+      );
+
+      console.log(`Activation Transaction ID: ${txData?.stageTransaction}`);
+
+      if (!txData?.stageTransaction) {
+        return {
+          result: false,
+          error: {
+            error: new Error("tx staging failed"),
+            step,
+          },
+        };
+      }
+
+      return {
+        result: true,
+        txId: txData.stageTransaction,
+      };
+    } catch (e: unknown) {
```

Comment:
If you're using unknown as type of Error, I hope there's type assertion in catch such as
```
  if (error instanceof Error) {
    errorMessage = error.message;
  }
```

Reply:
I used that to propagate error to the parent without modification.

I know that I can assert the type but what we should do if the assertion fail?
```typescript
catch (error: unknown) {
    if (error instance of Error) {
        return {
            result: false,
            error: {
                error,
                step
            }
        }
    }
    // What should we do here?
    // Doing nothing results the function return signiture to be ActivationResult | undefined
}
```

Reply:
Maybe...?
```typescript
catch (error: unknown) {
    if (error instance of Error) {
        return {
            result: false,
            error: {
                error,
                step
            }
        }
    }
    return {
        result: false,
        error: {
            error: new Error("Unknown error occurred"),
            step
        }
    }
}
```

Reply:
I guess just re-throwing might be simple answer for this unusual case?
https://stackoverflow.com/a/67160588

Reply:
It seems a good way, applied on [053492c](https://github.com/planetarium/9c-launcher/pull/2110/commits/053492c781882a9d4ec624d01190fcc635d5187d)